ocaml/xapi/xapi_vm.ml

```diff
@@ -254,9 +294,26 @@ let hard_shutdown ~__context ~vm =
 		(* Whether or not that worked, forget about the VDI *)
 		Db.VM.set_suspend_VDI ~__context ~self:vm ~value:Ref.null;
 		Xapi_vm_lifecycle.force_state_reset ~__context ~self:vm ~value:`Halted;
-	end else
-	Xapi_xenops.shutdown ~__context ~self:vm None;
-	Xapi_vm_helpers.shutdown_delay ~__context ~vm
+	end
+	| `Running
+	| `Paused ->
+		Xapi_xenops.shutdown ~__context ~self:vm None;
+		Xapi_vm_helpers.shutdown_delay ~__context ~vm
+	| `Halted -> ()
+
+let hard_reboot ~__context ~vm =
+	cancel ~__context ~vm ~ops:[ `clean_shutdown; `clean_reboot; `pool_migrate; `call_plugin; `suspend ];
+	(* Cancelling operations can cause the VM to now be shutdown *)
+	begin
+	match Db.VM.get_power_state ~__context ~self:vm with
+	| `Running
+	| `Paused ->
+		Xapi_xenops.reboot ~__context ~self:vm None
+	| `Halted ->
+		start ~__context ~vm ~start_paused:false ~force:false
```

Comment:
No, it was running when the call started (it got passed the allowed operations check), and very likely is only stopped because of the cancellation. This is the same behaviour as we have currently.